File: Server/services/Promotion.service.js
import PromotionRepository from '../repositories/Promotion.repository.js';
import { v2 as cloudinary } from 'cloudinary';
const PromotionServices = {
  async countPromotion() {
    return await PromotionRepository.countPromotion();
  },
  async changeIsDisable(id, isDisable) {
    return await PromotionRepository.changeIsDisable(id, isDisable);
  },
  async create(
    name,
    description,
    discount,
    start_date,
    end_date,
    quantity,
    code,
    file
  ) {
    if (
      name == null ||
      name == '' ||
      description == null ||
      description == '' ||
      discount == null ||
      discount == '' ||
      start_date == null ||
      start_date == '' ||
      end_date == null ||
      end_date == '' ||
      quantity == null ||
      quantity == '' ||
      code == null ||
      code == '' ||
      file == ''
    ) {
      cloudinary.uploader.destroy(file.filename);
      return false;
    } else {
      return await PromotionRepository.create(
        name,
        description,
        discount,
        start_date,
        end_date,
        file.path,
        quantity,
        code
      );
    }
  },
  async updateWithoutImage(
    id,
    name,
    description,
    discount,
    start_date,
    end_date,
    quantity,
    code
  ) {
    return await PromotionRepository.updateWithoutImage(
      id,
      name,
      description,
      discount,
      start_date,
      end_date,
      quantity,
      code
    );
  },
  async update(
    id,
    name,
    description,
    discount,
    start_date,
    end_date,
    image,
    quantity,
    code
  ) {
    if (
      name == null ||
      name == '' ||
      description == null ||
      description == '' ||
      discount == null ||
      discount == '' ||
      start_date == null ||
      start_date == '' ||
      end_date == null ||
      end_date == '' ||
      quantity == null ||
      quantity == '' ||
      code == null ||
      code == '' ||
      image == ''
    ) {
      cloudinary.uploader.destroy(image.filename);
      return false;
    } else {
      return await PromotionRepository.update(
        id,
        name,
        description,
        discount,
        start_date,
        end_date,
        image.path,
        quantity,
        code
      );
    }
  },
  async delete(id) {
    return await PromotionRepository.delete(id);
  },
  async findOne(id) {
    return await PromotionRepository.findOneByID(id);
  },
  async findAll() {
    return await PromotionRepository.findAll();
  },
  // Cần method gì thì tự implements !!
};

export default PromotionServices;

File: App User/app/src/main/java/com/example/myapplication/Admin/services/PromotionService.kt
package com.example.myapplication.Admin.services

import com.example.myapplication.Admin.modals.Promotion
import com.example.myapplication.Admin.modals.countPromotion
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.http.*


interface PromotionService {
    @GET("promotion/")
    suspend fun getAllPromotion(): List<Promotion>

    @GET("promotion/{id}")
    suspend fun getPromotionById(@Path("id") id: Int): Promotion

    @Multipart
    @POST("promotion/create")
    suspend fun createPromotion(
        @Part("name") name: RequestBody,
        @Part("description") description: RequestBody,
        @Part("discount") discount: RequestBody,
        @Part("start_date") start_date: RequestBody,
        @Part("end_date") end_date: RequestBody,
        @Part("quantity") quantity: RequestBody,
        @Part("code") code: RequestBody,
        @Part image: MultipartBody.Part
    ): Promotion

    @Multipart
    @PUT("promotion/update/{id}")
    suspend fun updatePromotion(
        @Path("id") id: Int,
        @Part("name") name: RequestBody,
        @Part("description") description: RequestBody,
        @Part("discount") discount: RequestBody,
        @Part("start_date") start_date: RequestBody,
        @Part("end_date") end_date: RequestBody,
        @Part("quantity") quantity: RequestBody,
        @Part("code") code: RequestBody,
        @Part image: MultipartBody.Part
    ): Promotion

    @PUT("promotion/updateWithoutImage/{id}")
    suspend fun updatePromotionWithoutImage(
        @Path("id") id: Int,
        @Body promotion: Promotion
    ): Promotion

    @DELETE("promotion/delete/{id}")
    suspend fun deletePromotion(@Path("id") id: Int): Boolean

    @PUT("promotion/disable/{id}")
    suspend fun disablePromotion(@Path("id") id: Int): Boolean

    @PUT("promotion/enable/{id}")
    suspend fun enablePromotion(@Path("id") id: Int): Boolean

    @GET("promotion/count")
    suspend fun countPromotion(): countPromotion
}